File: quant/app_manifest.py
"""iOS / App 功能清单 · 三分类精简结构（期权策略 / 做多票 / 空多双杀）。"""

from __future__ import annotations

import json
from datetime import datetime
from pathlib import Path
from typing import Any

from quant.strategy_catalog import (
    CORE_STRATEGY_IDS,
    build_strategy_audit,
    collect_strategy_snapshots,
    strategy_registry,
)

ROOT = Path(__file__).resolve().parents[1]
OUT_JSON = ROOT / "research" / "app_manifest.json"
IOS_JSON = ROOT / "ios" / "Resources" / "app_manifest.json"

# 顶栏三分类（App 仅保留这三类）
THS_CATEGORIES: list[dict[str, str]] = [
    {"id": "options", "name": "期权策略", "icon": "chart.line.uptrend.xyaxis", "color": "#10B981"},
    {"id": "long", "name": "做多票", "icon": "flame.fill", "color": "#E93030"},
    {"id": "short", "name": "空多双杀", "icon": "arrow.up.arrow.down.circle.fill", "color": "#A855F7"},
]

# 额外 feature（非 daily_pick 核心策略，但需在三分类中展示）
LAB_FEATURES: list[dict[str, Any]] = [
    {
        "id": "short_squeeze",
        "name": "安飞士做空",
        "category": "空多双杀",
        "ths_category": "short",
        "icon": "arrow.up.arrow.down.circle.fill",
        "script": "short_squeeze_daily.py",
        "config": "blowoff_short_config.json",
        "today_json": "research/short_squeeze_today.json",
        "description": "安飞士/SPCE 式：暴涨乏力→破位大阴确认做空 · 合并过热分+投机池+涨幅榜做空",
        "integrated_in_daily_pick": False,
        "view_type": "json_generic",
        "launcher": "空多双杀_运行一次.command",
    },
]

_CATEGORY_TO_THS = {
    # 新三分类
    "期权策略": "options",
    "做多票": "long",
    "空多双杀": "short",
    # 兼容旧分类名（防止历史数据落到默认桶）
    "期权收入": "options",
    "期权": "options",
    "聚合": "hub",
    "动量": "long",
    "量价": "long",
    "规律": "long",
}

_VIEW_TYPE = {
    "daily_pick": "daily_pick",
    "gain15": "gain15",
    "extreme20": "extreme20",
    "gainer10": "json_generic",
    "capital_flow": "generic_picks",
    "meme_long": "meme",
    "flow_strategy": "generic_picks",
}


def _view_type_for(entry_id: str, today_json: str) -> str:
    if entry_id in _VIEW_TYPE:
        return _VIEW_TYPE[entry_id]
    if "gain15" in today_json:
        return "gain15"
    if "flow_daily" in today_json or "flow_strategy" in today_json:
        return "generic_picks"
    if "ticker_pattern" in today_json:
        return "meme"
    if today_json.endswith(".json"):
        return "json_generic"
    return "history_only"


def _icon_for(entry_id: str, category: str) -> str:
    icons = {
        "daily_pick": "star.circle.fill",
        "gain15": "flame.fill",
        "extreme20": "bolt.circle.fill",
        "longshort_combo": "arrow.up.arrow.down.circle.fill",
        "gainer10": "bolt.fill",
        "capital_flow": "arrow.left.arrow.right",
        "meme_long": "sparkles",
        "fleet_csp": "ferry.fill",
        "bear_call": "phone.arrow.down.left",
        "flow_strategy": "chart.bar.fill",
        "sndk_iron": "shield.fill",
        "vrp": "waveform.path.ecg",
        "whipsaw_short": "arrow.down.right.circle.fill",
        "short_squeeze": "arrow.up.arrow.down.circle.fill",
    }
    return icons.get(entry_id, "square.grid.2x2")


def build_feature_list(root: Path | None = None) -> list[dict[str, Any]]:
    root = root or ROOT
    snap_map = {r["id"]: r for r in collect_strategy_snapshots(root)}
    audit = build_strategy_audit(root)
    audit_map = {r["id"]: r for r in audit.get("rows", [])}
    seen: set[str] = set()
    features: list[dict[str, Any]] = []

    for s in strategy_registry():
        snap = snap_map.get(s.id, {})
        ar = audit_map.get(s.id, {})
        ths = _CATEGORY_TO_THS.get(s.category, "hub")
        feat = {
            "id": s.id,
            "name": s.name,
            "category": s.category,
            "ths_category": ths,
            "icon": _icon_for(s.id, s.category),
            "script": s.script,
            "config": s.config or "",
            "today_json": s.today_json or "",
            "today_csv": s.today_csv or "",
            "history_csv": s.history_csv or "",
            "description": s.description,
            "integrated_in_daily_pick": s.integrated_in_daily_pick,
            "daily_pick_module": s.daily_pick_module or "",
            "launcher": s.launcher or "",
            "view_type": _view_type_for(s.id, s.today_json or ""),
            "actionable": snap.get("可开仓", 0),
            "watching": snap.get("观望", 0),
            "total": snap.get("总条目", 0),
            "has_data": snap.get("今日有数据", False),
            "data_date": snap.get("数据日期", "—"),
            "trades": ar.get("trades"),
            "win_rate": ar.get("win_rate", s.win_rate),
            "ann_return": ar.get("ann_return"),
            "max_dd": ar.get("max_dd"),
            "sharpe": ar.get("sharpe", s.sharpe),
            "audit_rank": ar.get("audit_rank"),
            "audit_score": ar.get("audit_score"),
            "audit_tier": ar.get("audit_tier"),
            "audit_verdict": ar.get("audit_verdict"),
            "audit_action": ar.get("audit_action"),
            "is_core": s.id in CORE_STRATEGY_IDS or s.id == "daily_pick",
        }
        features.append(feat)
        seen.add(s.id)

    for extra in LAB_FEATURES:
        if extra["id"] in seen:
            continue
        feat = {**extra, "is_core": False}
        feat.setdefault("actionable", 0)
        feat.setdefault("watching", 0)
        feat.setdefault("total", 0)
        feat.setdefault("has_data", False)
        feat.setdefault("data_date", "—")
        features.append(feat)
        seen.add(extra["id"])

    return features


def build_app_manifest(root: Path | None = None) -> dict[str, Any]:
    root = root or ROOT
    features = build_feature_list(root)
    by_cat: dict[str, list[dict]] = {}
    for f in features:
        by_cat.setdefault(f["ths_category"], []).append(f)

    quick_ids = [
        "daily_pick",
        # 首页主推（金刚区置顶）
        "gainer10", "short_squeeze",
        # ① 期权策略
        "bear_call", "fleet_csp", "sndk_iron", "vrp",
        # ② 做多票
        "capital_flow", "flow_strategy", "extreme20", "meme_long",
        # ③ 空多双杀
        "whipsaw_short",
    ]
    quick_entries = [f for fid in quick_ids for f in features if f["id"] == fid]
    core_features = [f for f in features if f.get("is_core") and f["id"] != "daily_pick"]

    audit = build_strategy_audit(root)

    return {
        "version": "4.0",
        "updated": datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
        "app_name": "美股量化",
        "tagline": "三分类精选 · 期权策略 / 做多票 / 空多双杀",
        "core_strategy_ids": list(CORE_STRATEGY_IDS),
        "core_count": len(CORE_STRATEGY_IDS),
        "strategy_audit": audit,
        "categories": THS_CATEGORIES,
        "quick_entries": quick_entries,
        "core_features": core_features,
        "features": features,
        "features_by_category": by_cat,
        "json_base_hint": "http://{host}:8502/",
        "streamlit_hint": "http://{host}:8501/",
        "total_features": len(features),
        "with_json_feed": sum(1 for f in features if f.get("today_json")),
    }


def export_app_manifest(root: Path | None = None) -> Path:
    root = root or ROOT
    doc = build_app_manifest(root)
    OUT_JSON.write_text(json.dumps(doc, ensure_ascii=False, indent=2), encoding="utf-8")
    IOS_JSON.parent.mkdir(parents=True, exist_ok=True)
    IOS_JSON.write_text(json.dumps(doc, ensure_ascii=False, indent=2), encoding="utf-8")
    return OUT_JSON


if __name__ == "__main__":
    p = export_app_manifest()
    print(f"→ {p}")
    print(f"→ {IOS_JSON}")

File: ios/Sources/AppManifest.swift
import Foundation

struct AppManifest: Codable {
    let version: String?
    let updated: String?
    let appName: String?
    let tagline: String?
    let coreCount: Int?
    let categories: [ManifestCategory]
    let quickEntries: [ManifestFeature]
    let features: [ManifestFeature]
    let strategyAudit: StrategyAudit?
    let totalFeatures: Int?

    enum CodingKeys: String, CodingKey {
        case version, updated, tagline, categories, features
        case appName = "app_name"
        case coreCount = "core_count"
        case quickEntries = "quick_entries"
        case strategyAudit = "strategy_audit"
        case totalFeatures = "total_features"
    }
}

struct StrategyAudit: Codable, Hashable {
    let updated: String?
    let method: String?
    let strategyCount: Int?
    let rows: [StrategyAuditRow]?

    enum CodingKeys: String, CodingKey {
        case updated, method, rows
        case strategyCount = "strategy_count"
    }
}

struct StrategyAuditRow: Codable, Identifiable, Hashable {
    let id: String
    let name: String?
    let auditRank: Int?
    let auditScore: Double?
    let auditTier: String?
    let auditVerdict: String?
    let auditAction: String?

    enum CodingKeys: String, CodingKey {
        case id, name
        case auditRank = "audit_rank"
        case auditScore = "audit_score"
        case auditTier = "audit_tier"
        case auditVerdict = "audit_verdict"
        case auditAction = "audit_action"
    }
}

struct ManifestCategory: Codable, Identifiable, Hashable {
    let id: String
    let name: String
    let icon: String
    let color: String?
}

struct ManifestFeature: Codable, Identifiable, Hashable {
    let id: String
    let name: String
    let category: String
    let thsCategory: String
    let icon: String
    let script: String?
    let config: String?
    let todayJson: String?
    let todayCsv: String?
    let historyCsv: String?
    let description: String?
    let integratedInDailyPick: Bool?
    let dailyPickModule: String?
    let launcher: String?
    let viewType: String?
    let terminalTab: String?
    let actionable: Int?
    let watching: Int?
    let total: Int?
    let hasData: Bool?
    let dataDate: String?
    let trades: Int?
    let winRate: Double?
    let annReturn: Double?
    let maxDd: Double?
    let sharpe: Double?
    let auditRank: Int?
    let auditScore: Double?
    let auditTier: String?
    let auditVerdict: String?
    let auditAction: String?

    enum CodingKeys: String, CodingKey {
        case id, name, category, icon, script, config, description, launcher
        case thsCategory = "ths_category"
        case todayJson = "today_json"
        case todayCsv = "today_csv"
        case historyCsv = "history_csv"
        case integratedInDailyPick = "integrated_in_daily_pick"
        case dailyPickModule = "daily_pick_module"
        case viewType = "view_type"
        case terminalTab = "terminal_tab"
        case actionable, watching, total
        case hasData = "has_data"
        case dataDate = "data_date"
        case trades
        case winRate = "win_rate"
        case annReturn = "ann_return"
        case maxDd = "max_dd"
        case sharpe
        case auditRank = "audit_rank"
        case auditScore = "audit_score"
        case auditTier = "audit_tier"
        case auditVerdict = "audit_verdict"
        case auditAction = "audit_action"
    }

    var detail: String { description ?? "" }
    var integrated: Bool { integratedInDailyPick ?? false }
    var signalTotal: Int { total ?? 0 }
    var isTerminalOnly: Bool { viewType == "terminal_only" }
    var hasJsonFeed: Bool { !(todayJson ?? "").isEmpty }
}

@MainActor
final class ManifestLoader: ObservableObject {
    static let shared = ManifestLoader()

    @Published var manifest: AppManifest?
    @Published var isLoading = false
    @Published var errorMessage: String?
    @Published var loadedFrom: String?

    func reload() async {
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }

        if let gh = AppConfig.githubManifestURL, let remote = await loadFromURL(gh) {
            manifest = remote
            return
        }

        let paths = ["app_manifest.json"]
        for path in paths {
            if let url = AppSettings.shared.jsonURL(for: path) {
                if let remote = await loadFromURL(url) {
                    manifest = remote
                    return
                }
            }
        }

        manifest = Self.fallbackManifest
        errorMessage = "无法从云端加载功能清单，请检查网络"
    }

    private func loadFromURL(_ url: URL) async -> AppManifest? {
        let busted = AppConfig.cacheBustedURL(url) ?? url
        do {
            var req = URLRequest(url: busted)
            req.cachePolicy = .reloadIgnoringLocalCacheData
            req.timeoutInterval = AppConfig.requestTimeout(for: busted)
            let (data, resp) = try await URLSession.shared.data(for: req)
            guard let http = resp as? HTTPURLResponse, (200..<300).contains(http.statusCode) else { return nil }
            let decoded = try JSONDecoder().decode(AppManifest.self, from: data)
            loadedFrom = url.host == "raw.githubusercontent.com" ? "GitHub" : url.host
            return decoded
        } catch {
            return nil
        }
    }

    func features(in categoryId: String) -> [ManifestFeature] {
        (manifest?.features.filter { $0.thsCategory == categoryId } ?? [])
            .sorted {
                let l = $0.auditRank ?? Int.max
                let r = $1.auditRank ?? Int.max
                if l != r { return l < r }
                return $0.name < $1.name
            }
    }

    /// 内置兜底（与 Python app_manifest 同步的核心条目）
    static let fallbackManifest = AppManifest(
        version: "4.0",
        updated: nil,
        appName: "美股量化",
        tagline: "核心9策略 · 同花顺式原生入口",
        coreCount: 9,
        categories: [
            ManifestCategory(id: "hub", name: "聚合", icon: "star.circle.fill", color: "#E93030"),
            ManifestCategory(id: "momentum", name: "动量", icon: "flame.fill", color: "#FF6B00"),
            ManifestCategory(id: "flow", name: "量价", icon: "arrow.left.arrow.right", color: "#3B82F6"),
            ManifestCategory(id: "pattern", name: "规律", icon: "sparkles", color: "#A855F7"),
            ManifestCategory(id: "options", name: "期权", icon: "chart.line.uptrend.xyaxis", color: "#10B981"),
            ManifestCategory(id: "lab", name: "实验室", icon: "flask.fill", color: "#78716C"),
            ManifestCategory(id: "terminal", name: "终端", icon: "chart.xyaxis.line", color: "#0EA5E9"),
        ],
        quickEntries: [],
        features: [],
        strategyAudit: nil,
        totalFeatures: 0
    )
}
